lib/backend/k8s/syncer.go

```diff
@@ -280,18 +291,34 @@ func (syn *kubeSyncer) readFromKubernetesAPI() {
 	var event watch.Event
 	var kvp *model.KVPair
 	var opts metav1.ListOptions
-	var openWatchers []watch.Interface
+	var openWatchers = map[string]watch.Interface{}
 	closeWatchers := func() {
 		for _, w := range openWatchers {
 			log.WithField("watcher", w).Debug("Closing old watcher.")
 			w.Stop()
 		}
-		openWatchers = nil
+		openWatchers = map[string]watch.Interface{}
 	}
 
 	// Always perform an initial snapshot.
 	needsResync := true
 
+	watchEventContinue := func(event watch.Event, key string) bool {
+		if event.Type == watch.Error {
+			log.Warnf("Event requires resync: %+v", event)
+			needsResync = true
+			return true
+		} else if event.Object == nil {
+			log.Warnf("Need new %v watch: %+v", key, event)
+			w := openWatchers[key]
+			log.WithField("watcher", w).Debug("Closing old watcher.")
+			w.Stop()
```

Comment:
Sure, I can rework it in this way - but it will result in a lot of repeated pattern code, which is really what I was trying to avoid.